src/contractor/contractor.cpp

```diff
@@ -45,37 +49,162 @@
+struct Turn final
 {
-    std::size_t operator()(const std::tuple<OSMNodeID, OSMNodeID, OSMNodeID> &k) const noexcept
+    std::uint64_t from, via, to;
+    Turn() : from(0), via(0), to(0) {}
+    Turn(const std::uint64_t &from, const std::uint64_t &via, const std::uint64_t &to)
+        : from(from), via(via), to(to)
+    {
+    }
+    Turn(const OSMNodeID &from, const OSMNodeID &via, const OSMNodeID &to)
+        : from(static_cast<std::uint64_t>(from)), via(static_cast<std::uint64_t>(via)),
+          to(static_cast<std::uint64_t>(to))
+    {
+    }
+    bool operator<(const Turn &rhs) const
+    {
+        return std::tie(from, via, to) < std::tie(rhs.from, rhs.via, rhs.to);
+    }
+    bool operator==(const Turn &rhs) const
     {
-        std::size_t seed = 0;
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<0>(k)));
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<1>(k)));
-        boost::hash_combine(seed, static_cast<uint64_t>(std::get<2>(k)));
-        return seed;
+        return std::tie(from, via, to) == std::tie(rhs.from, rhs.via, rhs.to);
     }
 };
-}
 
-namespace
+struct PenaltySource final
 {
+    double penalty;
+    std::uint8_t source;
+};
+
 template <typename T> inline bool is_aligned(const void *pointer)
 {
     static_assert(sizeof(T) % alignof(T) == 0, "pointer can not be used as an array pointer");
     return reinterpret_cast<uintptr_t>(pointer) % alignof(T) == 0;
 }
+} // anon ns
+
+BOOST_FUSION_ADAPT_STRUCT(Segment, (std::uint64_t, from)(std::uint64_t, to))
+BOOST_FUSION_ADAPT_STRUCT(SpeedSource, (unsigned, speed))
+BOOST_FUSION_ADAPT_STRUCT(Turn, (std::uint64_t, from)(std::uint64_t, via)(std::uint64_t, to))
+BOOST_FUSION_ADAPT_STRUCT(PenaltySource, (double, penalty))
+
+// Function parses a list of CSV files using "key,value,comment" grammar.
+// Key and Value structures must be a model of Random Access Sequence.
+// Also the Value structure must have source member that will be filled
+// with the index of the csv_filenames array.
+// Function returns a lambda function that maps input Key to boost::optional<Value>
+template <typename Key, typename Value>
+auto ParseCSV(
+    const std::vector<std::string> &csv_filenames,
+    const qi::rule<boost::spirit::line_pos_iterator<boost::spirit::istream_iterator>, Key()> &key,
+    const qi::rule<boost::spirit::line_pos_iterator<boost::spirit::istream_iterator>, Value()>
+        &value)
+{
+    try
+    {
+        tbb::spin_mutex mutex;
+        std::vector<std::pair<Key, Value>> lookup;
+        tbb::parallel_for(
+            std::size_t{0},
+            csv_filenames.size(),
+            [&](const std::size_t idx) {
+                const auto file_id = idx + 1; // starts at one, zero means we assigned the weight
+                const auto &filename = csv_filenames[idx];
+
+                std::ifstream input_stream(filename, std::ios::binary);
+                input_stream.unsetf(std::ios::skipws);
```

Comment:
here `skipws` must be *unset*, otherwise `ifstream` eats all spaces and eols, so `line_pos_iterator` can not correctly count line numbers.